### PR TITLE
Fix S3900: Rule should not raise an issue when compared with IsNullOr…

### LINKED ISSUE #693: Fix S3900: Rule should not raise an issue when compared with IsNullOrWhitespace/IsNullOrEmpty
### Description

The rule gives a false positive, indicating that a string variable hasn't been tested for null befor its use. BUT the variable is tested, just not with an `x == null` but with a `string.IsNullOrWhiteSpace()` or with a `string.IsNullOrEmpty()`. 

### Repro steps
```
public void a(string b)
{
     if (string.IsNullOrWhiteSpace(b))  return;
     var c = b.Split('_');
     Console.WriteLine(c.Count());
}
```
### Expected behavior

It souldn't fire the rule because the string b is being checked for null before it's use.

### Actual behavior

False positive is reported on line `var c = b.Split('_');`, marking b as being used without previos check.

### Known workarounds

First check for null explicitly as in `if (b == null) return;` and THEN check for IsNullOrWhiteSpace, which effectively beats the whole purpose of the nice string function IsNullOrWhiteSpace

### Related information

* SonarC# Version: SonarLint for Visual Studio 3.4.0.1732
* Visual Studio Version : 15.2 (26430.16)
* If running through the Scanner for MSBuild, its version: N/A

### LINKED ISSUE #826: Fix S3900: Arguments of public methods should be validated against null should take guarding methods into account
A lot developers use Guarding to do their parameter checking.

The implementation that was once introduced by Microsoft itself (CA1062) supported this quite nicely by code that could look like the following:

``` C#
/// <summary>Supplies input parameter guarding.</summary>
public static class Guard
{
    /// <summary>Guards the parameter if not null, otherwise throws an argument (null) exception.</summary>
    /// <typeparam name="T">The type to guard; cannot be a structure.</typeparam>
    /// <param name="parameter">The parameter to guard.</param>
    /// <param name="paramName">The name of the parameter.</param>
    [DebuggerStepThrough]
    public static T NotNull<T>([ValidatedNotNull]T parameter, string paramName) where T : class
    {
        if (ReferenceEquals(parameter, null))
        {
            throw new ArgumentNullException(paramName);
        }
        return parameter;
    }
}

/// <summary>Marks the NotNull argument as being validated for not being null, to satisfy the static code analysis.</summary>
/// <remarks>
/// Notice that it does not matter what this attribute does, as long as
/// it is named ValidatedNotNullAttribute.
/// </remarks>
[AttributeUsage(AttributeTargets.Parameter)]
sealed class ValidatedNotNullAttribute : Attribute { }
```
So, if the parameter of the guarding method is decorated with an attribute with the name 'ValidatedNotNullAttribute' it is assumed to have done so.

Could this be added?

Also see https://esmithy.net/2011/03/15/suppressing-ca1062/

### PR DESCRIPTION
**Pull request by valhristov**
…Whitespace/IsNullOrEmpty

Fix #693 
Fix #826

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/InvocationVisitor.cs [hunk 1]
```diff
@@ -148,9 +148,11 @@ private ProgramState HandleInstanceEqualsCall()
             return SetConstraintOnValueEquals(equals, newProgramState);
         }
 
-        private static readonly ImmutableHashSet<string> IsNullMethodNames = ImmutableHashSet.Create(
+        private static readonly ISet<string> IsNullMethodNames = new HashSet<string>
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/InvocationVisitor.cs]: Standard perf improvement, consistent with the rest of the code

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodArgumentsShouldBeCheckedForNull.cs [hunk 2]
```diff
@@ -64,5 +64,26 @@ public void CompliantCases(bool b, object o1, object o2, object o3, object o4, E
 
             Program.staticField.ToString(); // Compliant
         }
+
+        public void MoreCompliantCases(string s1, string s2)
+        {
+            if (string.IsNullOrEmpty(s1))
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodArgumentsShouldBeCheckedForNull.cs]: Are there any tests for e.g. _!string.IsNullOrEmpty(s1)_ ?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/InvocationVisitor.cs [hunk 3]
```diff
@@ -148,9 +155,18 @@ private ProgramState HandleInstanceEqualsCall()
             return SetConstraintOnValueEquals(equals, newProgramState);
         }
 
-        private static readonly ImmutableHashSet<string> IsNullMethodNames = ImmutableHashSet.Create(
+        private ProgramState HandleNullValidationMethod(int validatedArgumentIndex, int invocationArgsCount) =>
+            programState
+                .PopValues(invocationArgsCount - validatedArgumentIndex - 1)
+                .PopValue(out var guardedArgumentValue)
+                .PopValues(validatedArgumentIndex)
+                .SetConstraint(guardedArgumentValue, ObjectConstraint.NotNull);
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/SymbolicExecution/InvocationVisitor.cs]: Set NotNull on the argument that has ValidatedNotNull attribute, ignore the rest

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/EnumerableExtensions.cs [hunk 4]
```diff
@@ -100,5 +100,11 @@ public static IEnumerable<T> WhereNotNull<T>(this IEnumerable<T> enumerable)
                 }
             }
         }
+
+        public static int IndexOf<T>(this IEnumerable<T> enumerable, Func<T, bool> predicate) =>
+            enumerable
+                .Select((item, index) => new { item, index })
+                .FirstOrDefault(x => predicate(x.item))
+                ?.index ?? -1;
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/EnumerableExtensions.cs]: I don't like this ugly code in InvocationVisitor, hence the extension method

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodArgumentsShouldBeCheckedForNull.cs (hunk 2 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodArgumentsShouldBeCheckedForNull.cs]: It should not be required, as the `!` is not part of the invocation. The negation is handled in a separate node of the exploded graph.

**Comment by valhristov** [comment]
In general the invalid syntax tests should not be required as the graph is walked instruction by instruction and if something is null, it should have been handled already.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodArgumentsShouldBeCheckedForNull.cs (hunk 2 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodArgumentsShouldBeCheckedForNull.cs]: I think it is still useful to add a test to be sure this is covered. If not it's probably not needed to cover it right now.

**Comment by michalb-sonar** [comment]
I added `if (string.IsNullOrEmpty()) { }` to line 782 in `ConditionEvaluatesToConstant.cs` test and got the following error:
`Message: Issue with message '' not expected on line 781`

Not sure if we care much to fix it, but at least documenting this in test would be good.

**Comment by valhristov** [comment]
@michalb-sonar are you sure you didn't mess with the secondary locations? The rule reports secondary location when there is unreachable code and usually "Issue with message '' not expected" happens when a secondary location is not configured correctly. Adding invalid code works in a separate test case...

**Comment by michalb-sonar** [comment]
@valhristov quite possible possible, I only had a very quick look - probably missed that test with invalid code.

**Pull request merged**